### ISSUE TITLE
Inconsistent generic type parameter inference

### ISSUE THREAD

**Issue by OliverKovacs** [microsoft/TypeScript#57191]
### 🔎 Search Terms

"generics", "inconsistent behavior", "incorrect inference"

### 🕗 Version & Regression Information

This is the behavior in every version I tried, and I reviewed the FAQ for entries about generics.

### ⏯ Playground Link

https://www.typescriptlang.org/play?ts=5.3.3#code/PTAEDcEMCcEtIHYBdQEYBcokAtYGdQB3Ae2gGsDICBTADwAdqBjJagEwCgQsBPR0PKlABeUAHJBYgNxcwSPtQEAmEeLxLpHWRBjxkoJZhz5QbYtTwIxKEuQ7z+g1dGqQzCADY9QAbQmoxAF0ZB0V1Z1d3L18JDWCtbiZiNmotUNAAZVUnAB9lGW1aUAB+YoEhADNSXkY8ewVQABEAHloAPlUiulYENgIAV17qCtgEdhLy0ExaAorBllhiBFAK5ppaHr7M0DzBlJGxtjaAClpizBa8NoBKUABvAF8tCuOEfo8PUCpQQbIEYkIy2+uWU1ykQA

### 💻 Code

```ts
// version 1: this works as expected
// type s1 = 's1';
// type s2 = 's2';

// version 2: this doesn't work
type s1 = readonly ['s1'];
type s2 = readonly ['s2'];

// code

type S = s1 | s2;

// x ?? s1 for types
type D<x> = x extends undefined ? s1 : x;

function f<s extends S | undefined>(x?: D<s>) {}

f(null as unknown as s1 | s2);
```

### 🙁 Actual behavior

The generic type parameter of `f` on line 18 are reported as
```ts
f<s1>(x?: s1 | undefined)
```
when hovered, which then results in the error
```
Argument of type 'S' is not assignable to parameter of type 's1 | undefined'.
  Type 's2' is not assignable to type 's1'.
    Type '"s2"' is not assignable to type '"s1"'.(2345)
```
This is noteworthy, because if variant 1 is used (see [Playground Link](https://www.typescriptlang.org/play?ts=5.3.3#code/PTAEDcEMCcEtIHYBdQEYBcokAtYGdQB3Ae2gGsDICBTADwAdqBjJagEwCgkBPR0PVKAC8oAOQDRAbi69q-AEzCxeeVI4cQEGPGSh5mHPlBti1PAlEoS5DWB58BS6NUgmEAG26gA2uNSiAXWlNezkVJxc3Tx9xVSD1TSZiNmp1UNAAZSVHAB8FaVtQWlAAfhL+QQAzUixZPBk+ABEAHloAPiViulYENgIAV17qStgEdlKK0ExaAsrBllhiBFBK5ppaHr7M0DzBlJGxtjaACloSzBa8NoBKUABvAF91SuOEfvd3UCpQQbIEYkIy2+uQU10kQA)) the generic parameters are correctly inferred as
```ts
f<"s1" | "s2">(x?: "s1" | "s2" | undefined)
```

### 🙂 Expected behavior

This code snippet shouldn't result in an error, with the generic parameters of `f` correctly inferred as
```ts
f<s1 | s2>(x?: s1 | s2 | undefined)
```
At the very least I would expect variant 1 and variant 2 to produce the same result.

### Additional information about the issue

This is an aggressively simplified version of a problem encountered in actualy code. 

It is further interesting that making the argument `x` non-optional be removing the `?` seems to fix the issue.
However this is not a sufficient solution in general because sometimes (as in my case) `f()` without an argument is expected to be valid.

If this is correct behavior I would like to know why variant 1 and 2 are treated differently.

**Comment by RyanCavanaugh** [comment]
Generic inference will not produce a union in the presence of multiple type candidates unless they come from the same literal primitive family. This difference is intentional and on purpose.

I think it's kind of weird that we consider two different inference sites from the same location to be different candidates, though. Simpler version of that:
```ts
// Setup
type Box<T> = { value: T };
declare function foo<T>(arg: Box<T>): void;

declare const b1: Box<string | number>;
// OK, T = string | number
foo(b1);

declare const b2: Box<string> | Box<number>;
// Error, T = string
foo(b2);
```
but I'm not sure how to rectify that with the current algorithm without breaking something else. Will think more on it.

**Comment by fatcerberus** [comment]
> This difference is intentional and on purpose.

As opposed to intentional, but accidental. :wink:

On a more serious note, would it ever make sense to infer `T = string | number` from an inference candidate of `Box<string> | Box<number>`?  That would imply a parameter of type `Box<string | number>` for ’foo`, which doesn’t square with the inference source.  `T = string` is more arbitrary, but also feels more conceptually sound.
